src/graph/meter_roles.rs

```diff
@@ -0,0 +1,434 @@
+    }
+
+    fn with_multiple_grid_meters() -> (Vec<TestComponent>, Vec<TestConnection>) {
+        let (mut components, mut connections) = nodes_and_edges();
+
+        // Add a meter to the grid without successors
+        components.push(TestComponent(19, ComponentCategory::Meter));
+        connections.push(TestConnection::new(1, 19));
+
+        // Add a meter to the grid that has a battery meter and a PV meter as
+        // successors.
+        components.push(TestComponent(20, ComponentCategory::Meter));
+        connections.push(TestConnection::new(1, 20));
+
+        // battery chain
+        components.push(TestComponent(21, ComponentCategory::Meter));
+        components.push(TestComponent(
+            22,
+            ComponentCategory::Inverter(InverterType::Battery),
+        ));
+        components.push(TestComponent(23, ComponentCategory::Battery));
+        connections.push(TestConnection::new(20, 21));
+        connections.push(TestConnection::new(21, 22));
+        connections.push(TestConnection::new(22, 23));
+
+        // pv chain
+        components.push(TestComponent(24, ComponentCategory::Meter));
+        components.push(TestComponent(
+            25,
+            ComponentCategory::Inverter(InverterType::Solar),
+        ));
+        connections.push(TestConnection::new(20, 24));
+        connections.push(TestConnection::new(24, 25));
+
+        (components, connections)
+    }
+
+    fn without_grid_meters() -> (Vec<TestComponent>, Vec<TestConnection>) {
+        let (mut components, mut connections) = nodes_and_edges();
+
+        // Add an EV charger meter to the grid, then none of the meters
+        // connected to the grid should be detected as grid meters.
+        components.push(TestComponent(20, ComponentCategory::Meter));
+        components.push(TestComponent(21, ComponentCategory::EvCharger));
+        connections.push(TestConnection::new(1, 20));
+        connections.push(TestConnection::new(20, 21));
+
+        (components, connections)
+    }
+
+    fn assert_meter_role(
+        components: Vec<TestComponent>,
+        connections: Vec<TestConnection>,
+        filter: impl Fn(&ComponentGraph<TestComponent, TestConnection>, u64) -> Result<bool, Error>,
+        expected_grid_meters: Vec<u64>,
+    ) -> Result<(), Error> {
+        let graph = ComponentGraph::try_new(components.clone(), connections.clone())?;
+
+        let mut grid_meters = vec![];
+        for comp in graph.components() {
+            if filter(&graph, comp.component_id())? {
+                grid_meters.push(comp.component_id());
+            }
+        }
+        assert_eq!(grid_meters, expected_grid_meters);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_is_grid_meter() -> Result<(), Error> {
+        let (components, connections) = nodes_and_edges();
+        assert_meter_role(
+            components,
+            connections,
+            ComponentGraph::is_grid_meter,
+            vec![2],
+        )?;
```

Comment:
If the assertion in `assert_meter_role` fails, the error will show line 285, and not thisline, This would make the error difficult to identify.

I would recommend returning a result an error in line 285 instead.

Reply:
true, done.